### ISSUE TITLE
CodeGen bug?

### ISSUE THREAD

**Issue by LEI-Hongfaan** [dotnet/runtime#38555]
### Description

```csharp
using SharpDX;
using System;
class Program {
    static void Main(string[] args) {
        // Matrix comes from SharpDX.Mathematics
        Matrix matrixInvertible = new Matrix() {
            M11 = 0.1f, M22 = 0.2f, M33 = 0.5f,
            M41 = 10f, M42 = 20f, M43 = 30f, M44 = 1f, };
        // Matrix.Invert returns Matrix.Zero when input is not invertible
        Matrix result_0_OK = Matrix.Invert(matrixInvertible);
        // if (Math.Abs(det) == 0.0f)
        // ...
        // 00007FFED96E3D5E  vxorps      xmm1,xmm1,xmm1  
        // 00007FFED96E3D62  vucomiss    xmm0,xmm1  
        // 00007FFED96E3D66  jp          SharpDX.Matrix.Invert(SharpDX.Matrix ByRef, SharpDX.Matrix ByRef)+02FCh (07FFED96E3DBCh)  
        // 00007FFED96E3D68  jne         SharpDX.Matrix.Invert(SharpDX.Matrix ByRef, SharpDX.Matrix ByRef)+02FCh (07FFED96E3DBCh)  
        // 00007FFED96E3D6A  mov         rcx,7FFED97A8D30h  
        // 00007FFED96E3D74  mov         edx,18h  
        // 00007FFED96E3D79  call        CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE (07FFF392AF320h)  
        // ...
        for (int i = 0; 0x800000 > i; ++i) {
            Matrix.Invert(matrixInvertible);
        }
        // Matrix.Invert re-jitted.
        // if (Math.Abs(det) == 0.0f)
        // ...
        // 00007FFED96E4BB9  vxorps      xmm0,xmm0,xmm0  
        // 00007FFED96E4BBD  vmovss      dword ptr [rsp+88h],xmm0  
        // 00007FFED96E4BC6  vmovss      xmm0,dword ptr [rsp+8Ch]  
        // 00007FFED96E4BCF  vucomiss    xmm0,dword ptr [rsp+88h]  
        // 00007FFED96E4BD8  jp          SharpDX.Matrix.Invert(SharpDX.Matrix ByRef, SharpDX.Matrix ByRef)+0339h (07FFED96E4C89h)  
        // 00007FFED96E4BDE  mov         rcx,7FFED97A8D30h  
        // 00007FFED96E4BE8  mov         edx,18h  
        // 00007FFED96E4BED  call        CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE (07FFF392AF320h)  
        // ...
        // Expected:
        // vucomiss    xmm0,dword ptr [rsp+88h]  
        // jp          ...
        // jne         ...
        Matrix result_1_FAIL = Matrix.Invert(matrixInvertible);
        Console.WriteLine(result_0_OK.M11);
        // result_1_OK.M11 =~ 10
        Console.WriteLine(result_0_OK.Equals(result_1_FAIL));
        // False
        // Expected: True
        Console.WriteLine(result_1_FAIL.M11);
        // result_3_FAIL.M11 == 0.0f
        // Expected: result_1_OK.M11 =~ 10
    }
}
```

### Configuration

.NET SDK 5.0.100-preview.6.20318.15
Windows 10 x64

### Regression?

The code works fine with .NET Core SDK 3.1.105.

### Other information

The problem does NOT occur if I copy-paste the source code from Matrix.cs and compile it into a simplified library.

**Comment by sandreenko** [comment]
Hi @LEI-Hongfaan, thanks for the report, this issue was fixed by https://github.com/dotnet/runtime/pull/37379 (was not able to repro locally, so did iterations to find when it was fixed), so preview7 should not have it.

However, could you tell me which command did you use to compile this project? 
I was trying `dotnet run Test38555.csproj` with 
`<TargetFramework>netcoreapp5.0.0-preview</TargetFramework>`
and
```
  <ItemGroup>
    <PackageReference Include="SharpDX" Version="4.2.0" />
    <PackageReference Include="SharpDX.Mathematics" Version="4.2.0" />
  </ItemGroup>
```
in `.csproj` file, but got `error NU1202: Package SharpDX 4.2.0 is not compatible with net5.0-preview`. 

Could you write your repro steps for me please?

> The problem does NOT occur if I copy-paste the source code from Matrix.cs and compile it into a simplified library.
Probably inlining happens differently in that case.

**Comment by LEI-Hongfaan** [comment]
@sandreenko Thanks for your comment.

> in .csproj file, but got error NU1202: Package SharpDX 4.2.0 is not compatible with net5.0-preview.

Packages can be restored with `<TargetFramework>net5.0</TargetFramework>`. 

> this issue was fixed by #37379 

I compiled and ran my test project with daily build 5.0.100-preview.8.20351.5 (4938de92) and found no problem. I think you're right. However, I can not verify the fix against my original project because the newer preview has other problems.
